agreement/abstractions.go

```diff
@@ -72,19 +65,40 @@ var ErrAssembleBlockRoundStale = errors.New("requested round for AssembleBlock i
 // An BlockFactory produces an Block which is suitable for proposal for a given
 // Round.
 type BlockFactory interface {
-	// AssembleBlock produces a new ValidatedBlock which is suitable for proposal
+	// AssembleBlock produces a new AssembledBlock which is suitable for proposal
 	// at a given Round.
 	//
-	// AssembleBlock should produce a ValidatedBlock for which the corresponding
+	// AssembleBlock should produce a block for which the corresponding
 	// BlockValidator validates (i.e. for which BlockValidator.Validate
 	// returns true). If an insufficient number of nodes can assemble valid
 	// entries, the agreement protocol may lose liveness.
 	//
 	// AssembleBlock may return an error if the BlockFactory is unable to
-	// produce a ValidatedBlock for the given round. If an insufficient number of
+	// produce a AssembledBlock for the given round. If an insufficient number of
 	// nodes on the network can assemble entries, the agreement protocol may
 	// lose liveness.
-	AssembleBlock(basics.Round) (ValidatedBlock, error)
+	AssembleBlock(rnd basics.Round, partAddresses []basics.Address) (AssembledBlock, error)
+}
+
+// An AssembledBlock represents a Block produced by a BlockFactory
+// and must be finalized before being proposed by agreement.
+type AssembledBlock interface {
+	// WithSeed creates a copy of this AssembledBlock with its
+	// cryptographically random seed set to the given value.
+	//
+	// Calls to Seed() or to Digest() on the copy's Block must
+	// reflect the value of the new seed.
+	FinalizeBlock(seed committee.Seed, proposer basics.Address) ProposableBlock
+
+	Round() basics.Round
```

Comment:
Why do we need Round() now?

AssembledBlocks can't give out their Block()?

Reply:
It does seem good that an assembledBlock will no longer implement ValidatedBlock.